dev-6.1: функция-архиватор

lonelyfox-dev opened the pull request:
- НЕ ФИНАЛЬНАЯ ВЕРСИЯ. Вроде как работает, но возможно, нужны подкрутки, подстройка возвращаемых значений и т.д. Чтоб ничего лишнего не пердело и не свистело

- Комментов не много, соответственно. Пример работы для наглядности в директории lab1_misc

- Внутри lab1_misc: text.txt - архив, tree.txt - дерево архивируемой директории. При надобности, могу отправить саму директорию на репку

lonelyfox-dev committed:
Сделал функцию-архиватор

-----

- НЕ ФИНАЛЬНАЯ ВЕРСИЯ. Вроде как работает, но возможно, нужны подкрутки, подстройка возвращаемых значений и т.д. Чтоб ничего лишнего не пердело и не свистело

- Комментов не много, соответственно. Пример работы для наглядности в директории lab1_misc

- Внутри lab1_misc: text.txt - архив, tree.txt - дерево архивируемой директории. При надобности, могу отправить саму директорию на репку

lonelyfox-dev commented on lab1_misc/tree.txt:
```diff
@@ -0,0 +1,14 @@
+│   │   └── anotherfile.txt
+│   └── file.txt
+├── somefile.txt
+└── tree.txt
```
На этот файл не смотрим, забыл его вырезать

lonelyfox-dev replied:
(Который tree.txt)

SHTYKOVYCH replied:
Так бахни git rm --cashed <filename>

lonelyfox-dev committed:
Добавил тестовую директорию, поправил tree.txt

SHTYKOVYCH commented on lab1/archive.c:
```diff
@@ -0,0 +1,103 @@
+    while ((entry = readdir(dp)) != NULL)
+    {
+        lstat(entry->d_name, &statbuf);
+        if (entry->d_type == 4)
```
без комментиков тяжко конечно

SHTYKOVYCH commented on lab1/archive.c:
```diff
@@ -0,0 +1,103 @@
+        lstat(entry->d_name, &statbuf);
+        if (entry->d_type == 4)
+        {
+            if (strcmp(entry->d_name, ".") == 0 || strcmp(entry->d_name, "..") == 0 ) 
```
О нет, единство код стайла сравнения строк...

SHTYKOVYCH commented on lab1/archive.c:
```diff
@@ -0,0 +1,103 @@
+
+    int count = 0;
+
+    char *buff = malloc(sizeof(char) * 65536);
```
может это в стаке бахнуть?

SHTYKOVYCH commented on lab1/archive.c:
```diff
@@ -0,0 +1,103 @@
+    closedir(dp);   
+}
+
+void archive(char *dir)	// Требуется указывать полный путь до архивируемой папки. К примеру, /home/nikolay/Testdir
```
Относительные не работают?

SHTYKOVYCH replied:
И надо ли передавать в саму прогу абсолютный путь

SHTYKOVYCH commented on lab1/archive.c:
```diff
@@ -0,0 +1,103 @@
+
+    char *buff = malloc(sizeof(char) * 65536);
+    memset(buff, 0, 65536);
+    sprintDir(dir, buff, &count);
```
а нахера, я ж писал, что храним число записей в архиве как unsigned int. похер, что int будет, главное что сыро храним

SHTYKOVYCH commented on lab1/archive.c:
```diff
@@ -0,0 +1,103 @@
+
+    writeFile(filedescr, count_str, strlen(count_str));
+    seekFile(filedescr, strlen(count_str));
+    writeFile(filedescr, buff, strlen(buff));
```
тут, соответственно, будет просто write(filedescr, &count, 4)

SHTYKOVYCH commented on lab1/archive.c:
```diff
@@ -0,0 +1,103 @@
+    char count_str[12];
+    sprintf(count_str, "%d", count);
+
+    writeFile(filedescr, count_str, strlen(count_str));
```
тут вообще 12 байт запишется, а sizeof int вроде 4 считается

SHTYKOVYCH commented on lab1/archive.c:
```diff
@@ -0,0 +1,103 @@
+ 
+    if ((dp = opendir(dir)) == NULL)
+    {
+        perror(dir);
```
потом попробую над этим побольше подумать, может переведу все на такой механизм. Но это я без тебя просто вылизывать все буду

SHTYKOVYCH commented on lab1/archive.c:
```diff
@@ -0,0 +1,103 @@
+            char *buff = malloc(sizeof(char) * 1024);
+            memset(buff, 0, 1024);
+
+            sprintf(buff, "{name:\"%s\",parentDir:\"%s\",type=\"d\"}", entry->d_name, dir);
```
Я для этого написал jsonStringify

SHTYKOVYCH commented on lab1/archive.c:
```diff
@@ -0,0 +1,103 @@
+            strcat(str, meta);
+
+            int filedescr = open(abs_filename, O_RDONLY);
+            readFile(filedescr, contents, statbuf.st_size);
```
нахера читали чтение из файла, непонятно...
Бугурчу шо пиздец, понимаю, что не финал, но бугурчу

SHTYKOVYCH reviewed (changes requested):
Я пожалуй подожду финальной версии, пока есть недочеты
Структура файла в первых байтах точно неверная, так что надо менять. 
В общем, жду финальной версии

lonelyfox-dev committed:
Поправки к пулу по архиватору

-----

- Буфер ест не всё а по кусочкам

- json_stringify в деле

- Число сущностей пишется как надо (в случае чего, см. пример архива в lab1_misc/test.txt)

lonelyfox-dev commented on lab1/archive.c:
```diff
@@ -0,0 +1,121 @@
+
+    sprintDir(filedescr, dir, buff, jason, &count);
+
+    lseek(filedescr, 0L, SEEK_SET);
```
Если прям очень хочется пользоваться по максимуму написанными вспомогательными функциями, могу дописать функцию перевода указателя, чтоб можно было передавать третий аргумент

SHTYKOVYCH commented on lab1/archive.c:
```diff
@@ -13,66 +13,82 @@
 
             (*count)++;
-            sprintDir(entry->d_name, str, count);
+            sprintDir(filedescr, entry->d_name, str, jason, count); // Ре курсия
```
Т е ты сначала пишешь всю инфу о дочерних для текущей директории вхождений, а только потом о самой директории?

lonelyfox-dev committed:
Исправил порядок записи файлов в архив

Commit:
dev-6.1 / Функция архиватор
	- Правил ошибки компиляции(хз как у тебя компилировалось)

SHTYKOVYCH committed:
Merge branch 'master' into dev-6.1
